rpcperms/interceptor.go

```diff
@@ -0,0 +1,256 @@
+
+	// We'll add the macaroon interceptors. If macaroons aren't disabled,
+	// then these interceptors will enforce macaroon authentication.
+	unaryInterceptor := r.macaroonUnaryServerInterceptor()
+	macUnaryInterceptors = append(macUnaryInterceptors, unaryInterceptor)
+
+	strmInterceptor := r.macaroonStreamServerInterceptor()
+	macStrmInterceptors = append(macStrmInterceptors, strmInterceptor)
+
+	// Get interceptors for Prometheus to gather gRPC performance metrics.
+	// If monitoring is disabled, GetPromInterceptors() will return empty
+	// slices.
+	promUnaryInterceptors, promStrmInterceptors :=
+		monitoring.GetPromInterceptors()
+
+	// Concatenate the slices of unary and stream interceptors respectively.
+	unaryInterceptors := append(macUnaryInterceptors, promUnaryInterceptors...)
+	strmInterceptors := append(macStrmInterceptors, promStrmInterceptors...)
+
+	// We'll also add our logging interceptors as well, so we can
+	// automatically log all errors that happen during RPC calls.
+	unaryInterceptors = append(
+		unaryInterceptors, errorLogUnaryServerInterceptor(r.rpcsLog),
+	)
+	strmInterceptors = append(
+		strmInterceptors, errorLogStreamServerInterceptor(r.rpcsLog),
+	)
+
+	// If any interceptors have been set up, add them to the server options.
+	var serverOpts []grpc.ServerOption
+	if len(unaryInterceptors) != 0 && len(strmInterceptors) != 0 {
+		chainedUnary := grpc_middleware.WithUnaryServerChain(
+			unaryInterceptors...,
+		)
+		chainedStream := grpc_middleware.WithStreamServerChain(
+			strmInterceptors...,
+		)
+		serverOpts = append(serverOpts, chainedUnary, chainedStream)
+	}
+
+	return serverOpts
+}
+
+// errorLogUnaryServerInterceptor is a simple UnaryServerInterceptor that will
+// automatically log any errors that occur when serving a client's unary
+// request.
+func errorLogUnaryServerInterceptor(logger btclog.Logger) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo,
+		handler grpc.UnaryHandler) (interface{}, error) {
+
+		resp, err := handler(ctx, req)
+		if err != nil {
+			// TODO(roasbeef): also log request details?
+			logger.Errorf("[%v]: %v", info.FullMethod, err)
+		}
+
+		return resp, err
+	}
+}
+
+// errorLogStreamServerInterceptor is a simple StreamServerInterceptor that
+// will log any errors that occur while processing a client or server streaming
+// RPC.
+func errorLogStreamServerInterceptor(logger btclog.Logger) grpc.StreamServerInterceptor {
+	return func(srv interface{}, ss grpc.ServerStream,
+		info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+
+		err := handler(srv, ss)
+		if err != nil {
+			logger.Errorf("[%v]: %v", info.FullMethod, err)
+		}
+
+		return err
+	}
+}
+
+// macaroonUnaryServerInterceptor is a GRPC interceptor that checks whether the
+// request is authorized by the included macaroons.
+func (r *RPCInterceptor) macaroonUnaryServerInterceptor() grpc.UnaryServerInterceptor {
```

Comment:
This method (and `macaroonStreamServerInterceptor`) need to be exported because they're currently used by loop, pool and faraday for their macaroon checker.

Reply:
Ah. I guess we would need it to still be defined on the `*Service` also?

Reply:
You mean the `MacaroonService`? Not necessarily, I can instantiate the `RPCInterceptor` directly and use its methods, so no problem there as long as they're exported.

Reply:
Pushed a commit to export them!